src/app/tests/integration/chip_im_initiator.cpp

```diff
@@ -316,11 +321,8 @@ int main(int argc, char * argv[])
             goto exit;
         }
 
-        // Wait for response until the Message interval.
-        while (!MessageIntervalExpired())
-        {
-            DriveIO();
-        }
+        // Suspend current thread and wait for response until gMessageInterval interval.
+        gCond.wait_for(lck, std::chrono::seconds(gMessageInterval));
```

Comment:
should we check for timeout vs no_timeout being returned here and below?

Reply:
Update the comments to make it clear, we should suspend current thread during gMessageIntervalSeconds, or until notified (if the latter happens first). 

If the current thread is resumed because of timeout, this will be catched later since gWaitingForCommandResp is not set.

        // Check if expected response was received.
        if (gWaitingForCommandResp)
        {
            printf("Invoke Command: No response received\n");
            gWaitingForCommandResp = false;
        }

Reply:
My comment was not because of lack of code understanding - I can read and see this timed wait-for.

My comment is that it is generally unusuall to wait with a timeout but not care if a timeout occured or not. It looks ok becuase gWaiting is used as a flag to figure out if the condition was notified or not, however I am unsure if this works correctly in the loop. It requires mental work for me to figure out if it is ok for code to just move ahead in both timeout and notify and if we fail to notify we will just have bad performance (too long wait) without ever noticing.

Please do a check and if the `wait_for` returns timeout, the code should error out. This will force us to have correct conditioning and may also require you to hold appropriate locks to ensure a sane concurrency model.

Reply:
Looking through the code it looks as if not receiving a response was never an error. Is this ok? When writing the above comment I believed we want to error out.

If this is not the case, that bit would be useful as a comment: why do we think its ok to not receive responses?

Reply:
Acked, using return value of wait_for to check if the result is not received because of timeout.

At last, it check if all responses are received, any missing will be treated as an error.
if (err != CHIP_NO_ERROR || (gCommandRespCount != kMaxCommandMessageCount))